### PR TITLE
Show page labels instead of page numbers

### PR DESCRIPTION
**Pull request by EricHanLiu**
See issue #434 

Also fixes v2 manifest parsing where manifest had values that were arrays of objects (instead of objects with values potentially being arrays (common in v3 manifests, already handled)) 

Will display page labels properly for each page in book view, and if the last page in book view is alone (even number of pages)

### CODE REVIEW

File: source/js/viewer-core.js [hunk 1]
```diff
@@ -1350,16 +1351,18 @@ export default class ViewerCore
      *
      * @param pageIndex
      */
-    setCurrentPage (pageIndex)
+    setCurrentPages (currentPage, visiblePages)
     {
-        if (this.viewerState.currentPageIndex !== pageIndex)
+        // these arrays will never contain objects so this should be safe
+        if (JSON.stringify(this.viewerState.currentPageIndices) !== JSON.stringify(visiblePages))
```
Comment (ahankinson) [inline source/js/viewer-core.js]: What is the purpose of passing this through `stringify`?

Comment (EricHanLiu) [reply source/js/viewer-core.js]: This seems to be the easiest way to compare simple arrays (that don't contain objects). Regular strict inequality doesn't work

**Comment by EricHanLiu** [comment]
Issue with the bottom strip of a page not being considered part of the page (so `isPageVisible` despite this portion being displayed will return false). 

I think it's because of the plugin icons, will test.

**Comment by EricHanLiu** [comment]
not because of the plugin icons, nor the page indices change (in this PR).

the issue is from `document-layout.js` with the page region dimensions being wrong. created a separate issue, see #436

**Comment by EricHanLiu** [comment]
Issue #436 fixed.

Still need to get rid of the `activePageIndex` and change all methods that depend on one page index to return a set of values corresponding to the array of pages.

**Comment by EricHanLiu** [comment]
By replacing the `activePageIndex`, if the user is sufficiently zoomed out such that multiple pages are visible there should be one that takes precedence over others when considering variables/methods that **shouldn't** return a set of values corresponding to the array of visible pages, but rather just a single value (such as [`pageOffset`](https://github.com/DDMAL/diva.js/blob/page-labels/source/js/diva.js#L388)).

In these cases, I think the best thing to do would be to get the middle page from the array of visible pages, since that is most likely the one the viewer is centred on. eg. getting the 0th indexed page can often correspond to a page that is 90% out of the viewport.

This only applies to methods/variables where it wouldn't make sense to use a set of values (such as the `p` state, which corresponds to the page index to `goDirectlyTo`. This shouldn't be set to multiple pages).  

Other methods like `getCurrentCanvas()` should become `getCurrentCanvases()`, which will return all the canvases for pages visible in the viewer.

Thoughts @ahankinson ?

**Comment by EricHanLiu** [comment]
Think it's best to keep the activePageIndex variable. Removing it would just remove a shorthand for essentially the midpoint of the visiblePages array for variables/functions that shouldn't be/return a set.

If zoomed out and 3 pages are visible, but only one full page is visible, then the active page should be that one page (and not the half-cutoff one above or below)

**Comment by ahankinson** [comment]
The page tools seem to not be properly aligned now?

**Comment by ahankinson** [comment]
image

**Comment by EricHanLiu** [comment]
Are you at the latest commit? They are aligned for me, I believe I fixed this with the last commit relating to issue #436

**Comment by ahankinson** [comment]
The page labels do not seem to change that the appropriate spot. I've made a movie showing this

https://www.dropbox.com/s/t2jsjxzjfs5j143/page-labels.mov?dl=0

**Comment by EricHanLiu** [comment]
@ahankinson they change when the page tool plugins come into view, since that is part of the page. The padding that's calculated (40px) for the plugins is defined as part of the page region (not my implementation).

The issue with #436  that had to be fixed was adding this same padding at the bottom of each page region so that the page didn't cut off prematurely. I believe it's always been this way where the plugins were included in the region though

**Comment by ahankinson** [comment]
I'm wondering if there's a better way to do this.

How about we try adding the page label to each page. Upper-right, opposite the page tools, add small text saying what the page label is. That would be the clearest, I think.

**Comment by EricHanLiu** [comment]
Added page labels to each page

**Comment by EricHanLiu** [comment]
Closing this in favor of PR #437 , which contains these changes as well.

The issue holding this PR back (#443) has been fixed.

**Pull request closed without merging**